method/DeepST/DeepST.py

```diff
@@ -0,0 +1,227 @@
+        )
+
+    if args.image is not None:
+        adata.uns["image"] = np.array(Image.open(args.image))
+    else:
+        adata.uns["image"] = None
+
+    return adata
+
+adata = get_anndata(args)
+adata.var_names_make_unique()
+
+#######
+# NOTE: Seed cannot be used here
+
+# First, set up DeepST environment
+import tempfile
+import os, sys
+import scanpy as sc
+import numpy as np
+import pandas as pd
+import scipy.sparse as sp
+import torch
+
+
+# Work in a temprary folder
+with tempfile.TemporaryDirectory() as tmpdir:
+    gitdir = f"{str(tmpdir)}/DeepST"
+
+    # Clone the repository to the specific commit
+    os.system(
+    f"""
+    git clone https://github.com/JiangBioLab/DeepST.git {gitdir}
+    cd {gitdir} 
+    git reset --hard 1daa513
+    """)
+
+    # Set working directory as deepST directory
+    sys.path.append(f"{gitdir}/deepst")
+    import adj
+    
+    # From DeepST import run
+    import importlib.util
+    # Import the main wrapper DeepST module
+    spec=importlib.util.spec_from_file_location("deepST", f"{gitdir}/deepst/DeepST.py")
+    deepST = importlib.util.module_from_spec(spec)
+    sys.modules["deepST"] = deepST
+    spec.loader.exec_module(deepST)
+    
+    # Set up deepST object
+    deepen = deepST.run(save_path = None,
+                 task = "Identify_Domain", 
+                 pre_epochs = config["pre_epochs"],  # change based on hardware
+                 epochs = config["epochs"],  # change based on hardware
+                 use_gpu = config["use_gpu"],
+                 )
+    
+    # adopted code from deepst/adj.py
+    """ Store original adjacency matrix (without diagonal entries) for later """
+    adj_pre = adata.obsp["spatial_connectivities"]
+    adj_pre = sp.coo_matrix(adj_pre)
+
+    adj_pre = adj_pre - sp.dia_matrix((adj_pre.diagonal()[np.newaxis, :], [0]), shape=adj_pre.shape)
+    adj_pre.eliminate_zeros()
+
+    """ Some preprocessing."""
+    # intiate a "graph" item, k is a placeholder doesn't matter
+    graph = adj.graph(data = adata, k = 1)
+    adj_norm = graph.pre_graph(adj_pre)
+    adj_label = adj_pre + sp.eye(adj_pre.shape[0])
+    adj_label = torch.FloatTensor(adj_label.toarray())
+    norm = adj_pre.shape[0] * adj_pre.shape[0] / float((adj_pre.shape[0] * adj_pre.shape[0] - adj_pre.sum()) * 2)
+
+    graph_dict = {
+                 "adj_norm": adj_norm,
+                 "adj_label": adj_label,
+                 "norm_value": norm }
+
+    # NOTE: There're so many hyperparameters but I would not change any...the last few commits the author made is to change some default parameters so I assume the author has figure out the "best" config for this model
+    
+    # Conv_type: ["GCNConv", "SAGEConv", "GraphConv", "GatedGraphConv", "ResGatedGraphConv", "TransformerConv", "TAGConv", "ARMAConv", "SGConv", "MFConv", "RGCNConv", "FeaStConv", "LEConv", "ClusterGCNConv"]
+    data = adata.obsm["reduced_dimensions"].astype(np.float64)
+    
+    deepst_embed = deepen._fit(data, 
```

Comment:
Is this part of the public API

Reply:
Do you mean the conv_type? They don't have ANY documentation on their model but in[ their script ](https://github.com/JiangBioLab/DeepST/blob/main/deepst/model.py#L63)they gave those options...

Reply:
No I meant the `._fit`, usually underscore would indicate private functions, but it seems their whole API is built like that

Reply:
Oh yes all the main functions are starting with a “_”. It should be public as they also use those functions in their github page example